### PR TITLE
Add not_null version of serialization tool

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Split out from #2613 

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]

File: src/Parallel/Serialize.hpp [hunk 1]
```diff
@@ -54,3 +53,18 @@ T deserialize(const void* const data) noexcept {  // NOLINT
   reader | result;
   return result;
 }
+
+/*!
+ * \ingroup ParallelGroup
+ * \brief Deserialize an object using PUP.
+ *
+ * \tparam T the type to deserialize to. The type to deserialize as must be
```
Comment (kidder) [inline src/Parallel/Serialize.hpp]: I don't think you need to explicitly specify the type

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by moxcodes** [comment]
fixup posted.
Cheers!

**Comment by nilsdeppe** [comment]
you have my permission to squash

**Review by kidder — COMMENTED** [review]
when squashing you can put back the first part of the comment i.e.`\tparam T the type to deserialize to.`

**Comment by moxcodes** [comment]
done.
Cheers!

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**